### PR TITLE
18 implement lazyt type for delayed evaluation with memoization

### PR DESCRIPTION
**Pull request by tovabry**
PR for issue #18. This issue is dependent on #6 Try<T>, and Try<T> is dependent on the Either class, so that could be confusing for the reviewers. What's supposed to be reviewed in this issue is the Lazy<T> class and Lazy<T> test class, not the other classes. 

## Summary by CodeRabbit

- **New Features**
  - Introduced a new lazy evaluation component that computes values on first access and caches them, enabling deferred computation along with transformation and filtering capabilities—all with thread-safe behavior.
- **Tests**
  - Added a thorough test suite covering typical usage, concurrent access, and edge cases such as handling nulls and error scenarios.

### CODE REVIEW

**Review by Oggeboii — CHANGES_REQUESTED** [review]
Apply "minor" label to PR

**Review by Oggeboii — DISMISSED** [review]

File: src/main/java/org/fungover/breeze/control/Lazy.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    public T get() {
+        if (!evaluated) {
+            synchronized (this) {
+                if (!evaluated) {
+                    value = supplier.get();
+                    evaluated = true;
+                }
+            }
+        }
+        return value;
```
Comment (AntonGrubisic) [inline src/main/java/org/fungover/breeze/control/Lazy.java:72]: Could use a AtomicReference in the class to store the value for safe updates to value.
Nothing needed though, just a thought.

Code looks good, signed!

**Review by AntonGrubisic — DISMISSED** [review]
Good code!
Merge ready, left a comment on the code.

**Review by Lhlmlund — CHANGES_REQUESTED** [review]
A very great foundation and implementation of lazyT type, well done with good evaluation and error handling, there are however a few minor improvements that could be added, also make sure to checkout CodeRabbit's suggestions for some nice tips!

Checkout public Try<T> filter(Predicate<? super T> predicate) in "Try.Java" 

The NoSuchElementException returns ("The value 'does satisfy' the predicate") instead of 
"The value does 'not' satisfy the predicate", just a typo but a meaningful one.

but **OBS**, that returning the NoSuchElementException to the current filter introduces another problem, by Throwing a new Exception, in this case a NoSuchElementException with the get method you are losing the original value of the failure, and because it overwrites it with your newely created one it can make it harder to debug and error handle it.

You could instead return the original error and then introduce the new one if it fails the predicate, like this:

public Try<T> filter(Predicate<? super T> predicate) {
    Objects.requireNonNull(predicate);
    if (isFailure()) return this; // Preserve the original failure
    return predicate.test(get()) 
        ? this 
        : failure(new NoSuchElementException("Value does not satisfy the predicate"));
}






Also I notice the use of " if(isSuccess()) { get(); } catch logic, I think it could be a nice helper method instead to improve readability and prevent code duplication!  

For example you could implement 


private <U> Try<U> executeSafely(Function<? super T, ? extends U> function) {
    if (isSuccess()) {
        try {
            return success(function.apply(get()));
        } catch (Throwable throwable) {
            return failure(throwable);
        }
    }
    return failure(((Failure<T>) this).throwable());
}

and in turn change flatMap map and recover 

public <U> Try<U> map(Function<? super T, ? extends U> mapper) {
    Objects.requireNonNull(mapper);
    return executeSafely(mapper);
}

public <U> Try<U> flatMap(Function<? super T, Try<U>> mapper) {
    Objects.requireNonNull(mapper);
    return executeSafely(value -> mapper.apply(value).get());
}

public Try<T> recover(Function<? super Throwable, ? extends T> recoverFunction) {
    Objects.requireNonNull(recoverFunction);
    return isSuccess() ? this : executeSafely(recoverFunction.compose(((Failure<T>) this)::throwable));
}




Also I would suggest to check for null values in your current constructor like so: 

private Lazy(Supplier<? extends T> supplier) {
    this.supplier = Objects.requireNonNull(supplier, "Supplier must not be null");
    this.value = null;
    this.evaluated = false;
}

this will prevent any potential NullPointerException

File: src/test/java/org/fungover/breeze/LazyTest.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        @Test
+        @DisplayName("filter value calculates only once")
+        void filterValueCalculatesOnlyOnce() {
+            AtomicInteger counter = new AtomicInteger(0);
+            Lazy<Optional<Integer>> lazyOptional = Lazy.of(() -> {
+                counter.incrementAndGet();
+                return 89;
+            }).filter(i -> i > 88);
+
+            lazyOptional.get();
```
Comment (Lhlmlund) [inline src/test/java/org/fungover/breeze/LazyTest.java:96]: You could remove this additional .get here

File: src/test/java/org/fungover/breeze/LazyTest.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+            Lazy<String> lazy = Lazy.of(() -> {
+                countedTimes.incrementAndGet();
+                return "computed value";
+            });
+
+            String value1 = lazy.get();
+            String value2 = lazy.get();
+
+            assertEquals(1, countedTimes.get());
+            assertNotEquals(2, countedTimes.get());
```
Comment (Lhlmlund) [inline src/test/java/org/fungover/breeze/LazyTest.java:50]: assertNotEquals(2, countedTimes.get()); is redundant here, assertEquals 1 is enough.
  
 essentially the test would already fail if assertEquals != 1

**Comment by tovabry** [comment]
Thanks for the reviews! I have made some changes, please take a look at them!
So this issue is a bit confusing to review, I merged the Try<T> class done by Erik (Lazy<T> is dependent on Try<T>), which makes it seem like I've made a pull request for the Try<T> class aswell as the Lazy<T> which is the issue I've been working on. But the classes that's supposed to be reviewed here is just Lazy and LazyTest. Sorry for the confusion! @Lhlmlund @Oggeboii @AntonGrubisic

File: src/main/java/org/fungover/breeze/control/Lazy.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        });
+    }
+
+    /**
+     * Converts the lazy value to an optional value.
+     *
+     * @return an optional value
+     */
+    public Optional<T> toOption() {
+        return isEvaluated() ? Optional.ofNullable(value) : Optional.empty();
```
Comment (Lhlmlund) [inline src/main/java/org/fungover/breeze/control/Lazy.java:126]: could add some further javadoc to toOption for more descriptive debugging

/**
 * Converts the lazy value to an optional, reflecting its evaluation state.
 * This method does not force evaluation; it returns Optional.empty() if the value has not been computed.
 *
 * @return an optional containing the value if already evaluated, otherwise Optional.empty()
 */

**Review by Lhlmlund — APPROVED** [review]
Lazy and Lazytest is very good, covering all the requirements met in the issue well, you have a strong test coverage and thread-safety mechanisms in place, overall very good imlementation. 

Keep an eye out for coderabbits suggestions to improve general performance

**Comment by tovabry** [comment]
Merge conflicts solved and some other small changes done

**Review by Oggeboii — APPROVED** [review]
Great job solving the requested changes by @Lhlmlund.  Very well documented with Javadoc.
The "Nested" tests creates good structure

**Review by Stelle83 — APPROVED** [review]
Overall, the changes look good and improve the thread safety and robustness of the code.

Thread Safety:
Replacing the array with AtomicInteger is a good change for ensuring thread safety when multiple threads are incrementing the count.
The use of calculationCount.incrementAndGet() provides atomicity and avoids potential race conditions.
Assertions:
Correctly updated the assertions to use calculationCount.get(), which is necessary after changing to AtomicInteger.
